File: Assets/Warlock/Scripts/Mvp/NetworkContract.cs
using System.Collections.Generic;

namespace Warlock.Mvp
{
    public static class NetworkContract
    {
        public static readonly IReadOnlyList<string> MatchEvents = new[]
        {
            "RoomCreated", "PlayerJoined", "RoomSettingsUpdated"
        };

        public static readonly IReadOnlyList<string> RoundEvents = new[]
        {
            "RoundStarted", "RoundEnded", "MatchEnded"
        };

        public static readonly IReadOnlyList<string> CombatEvents = new[]
        {
            "SkillCastRequested", "SkillCastConfirmed", "DamageApplied", "KnockbackApplied", "PlayerEliminated"
        };

        public static readonly IReadOnlyList<string> EconomyEvents = new[]
        {
            "RankPointsGranted", "GoldGranted", "ShopPurchaseRequested", "ShopPurchaseConfirmed", "ShopPurchaseRejected", "LoadoutUpdated"
        };

        public static readonly IReadOnlyList<string> ArenaEvents = new[]
        {
            "ShrinkStateUpdated", "BoundaryDotTick"
        };

        public static RuleResult ValidateHostSkillCastRequest(RoundState phase, CastValidationResult cooldownValidation)
        {
            if (phase != RoundState.Combat)
            {
                return RuleResult.Fail("ROUND_PHASE_INVALID");
            }

            if (!cooldownValidation.Ok)
            {
                return RuleResult.Fail(cooldownValidation.Reason ?? "COOLDOWN_VALIDATION_FAILED");
            }

            return RuleResult.Success();
        }

        public static RuleResult ValidateHostShopPurchase(RoundState phase, int playerGold, int price)
        {
            if (phase != RoundState.Shop)
            {
                return RuleResult.Fail("SHOP_PHASE_REQUIRED");
            }

            if (playerGold < 0)
            {
                return RuleResult.Fail("NEGATIVE_GOLD_INVALID");
            }

            if (price < 0)
            {
                return RuleResult.Fail("NEGATIVE_PRICE_INVALID");
            }

            if (playerGold < price)
            {
                return RuleResult.Fail("INSUFFICIENT_GOLD");
            }

            return RuleResult.Success();
        }
    }
}

File: src/network-contract.js
import { ROUND_STATES } from './constants.js';

export const NETWORK_EVENTS = Object.freeze({
  match: Object.freeze(['RoomCreated', 'PlayerJoined', 'RoomSettingsUpdated']),
  round: Object.freeze(['RoundStarted', 'RoundEnded', 'MatchEnded']),
  combat: Object.freeze([
    'SkillCastRequested',
    'SkillCastConfirmed',
    'DamageApplied',
    'KnockbackApplied',
    'PlayerEliminated',
  ]),
  economy: Object.freeze([
    'RankPointsGranted',
    'GoldGranted',
    'ShopPurchaseRequested',
    'ShopPurchaseConfirmed',
    'ShopPurchaseRejected',
    'LoadoutUpdated',
  ]),
  arena: Object.freeze(['ShrinkStateUpdated', 'BoundaryDotTick']),
});

export function validateHostSkillCastRequest({ phase, cooldownValidation }) {
  if (phase !== ROUND_STATES.COMBAT) {
    return { ok: false, reason: 'ROUND_PHASE_INVALID' };
  }

  if (!cooldownValidation.ok) {
    return { ok: false, reason: cooldownValidation.reason };
  }

  return { ok: true };
}

export function validateHostShopPurchase({ playerGold, price, phase }) {
  if (phase !== ROUND_STATES.SHOP) {
    return { ok: false, reason: 'SHOP_PHASE_REQUIRED' };
  }

  if (playerGold < 0) {
    return { ok: false, reason: 'NEGATIVE_GOLD_INVALID' };
  }

  if (price < 0) {
    return { ok: false, reason: 'NEGATIVE_PRICE_INVALID' };
  }

  if (playerGold < price) {
    return { ok: false, reason: 'INSUFFICIENT_GOLD' };
  }

  return { ok: true };
}
